connect-inject/container_init.go

```diff
@@ -72,11 +72,19 @@ func (h *Handler) containerInit(pod corev1.Pod, k8sNamespace string) (corev1.Con
 	}
 
 	// When ACLs are enabled, the ACL token returned from `consul login` is only
-	// valid for a service with the same name as the ServiceAccountName.
-	if data.AuthMethod != "" && data.ServiceName != pod.Spec.ServiceAccountName {
+	// valid for a service with the same name as the ServiceAccountName. If the
+	// annotation for service name is set, check against that.
+	if data.AuthMethod != "" && data.ServiceName != "" && data.ServiceName != pod.Spec.ServiceAccountName {
 		return corev1.Container{}, fmt.Errorf("serviceAccountName %q does not match service name %q", pod.Spec.ServiceAccountName, data.ServiceName)
 	}
 
+	// If the service name annotation does not exist, set the service account name flag contents for the init container
+	// command. If the init container sees the service account name flag as non-empty, it will check it against the
+	// Kubernetes service name to ensure they are the same.
```

Comment:
Note here and in the tests that if the service name annotation has *not* been set, the flag will be set as empty, which is meaningful in the connect-init command.